lib/node/localNode.go

```diff
@@ -0,0 +1,208 @@
+package sebaknode
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"boscoin.io/sebak/lib/common"
+
+	"github.com/stellar/go/keypair"
+)
+
+type NodeFromJSON struct {
+	Alias      string                `json:"alias"`
+	Address    string                `json:"address"`
+	Endpoint   *sebakcommon.Endpoint `json:"endpoint"`
+	Validators map[string]*Validator `json:"Validators"`
+	State      NodeState             `json:"state"`
+}
+
+type LocalNode struct {
+	sync.Mutex
+
+	keypair *keypair.Full
+
+	state      NodeState
+	alias      string
+	address    string
+	endpoint   *sebakcommon.Endpoint
+	validators map[ /* Node.Address() */ string]*Validator
+}
+
+func (n *LocalNode) String() string {
+	return n.Alias()
+}
+
+func (n *LocalNode) Equal(a Node) bool {
+	if n.Address() == a.Address() {
+		return true
+	}
+
+	return false
+}
+
+func (n *LocalNode) DeepEqual(a Node) bool {
+	if !n.Equal(a) {
+		return false
+	}
+	if n.Endpoint().String() != a.Endpoint().String() {
+		return false
+	}
+
+	return true
+}
+
+func (n *LocalNode) State() NodeState {
+	return n.state
+}
+
+func (n *LocalNode) SetBooting() {
+	n.state = NodeStateBOOTING
+}
+
+func (n *LocalNode) SetCatchup() {
+	n.state = NodeStateCATCHUP
+}
+
+func (n *LocalNode) SetConsensus() {
+	n.state = NodeStateCONSENSUS
+}
+
+func (n *LocalNode) SetTerminating() {
+	n.state = NodeStateTERMINATING
+}
+
+func (n *LocalNode) Address() string {
+	return n.address
+}
+
+func (n *LocalNode) Keypair() *keypair.Full {
+	return n.keypair
+}
+
+func (n *LocalNode) SetKeypair(kp *keypair.Full) {
+	n.address = kp.Address()
+	n.keypair = kp
+}
+
+func (n *LocalNode) Alias() string {
+	return n.alias
+}
+
+func (n *LocalNode) SetAlias(s string) {
+	n.alias = s
+}
+
+func (n *LocalNode) Endpoint() *sebakcommon.Endpoint {
+	return n.endpoint
+}
+
+func (n *LocalNode) HasValidators(address string) bool {
+	_, found := n.validators[address]
+	return found
+}
+
+func (n *LocalNode) GetValidators() map[string]*Validator {
+	return n.validators
+}
+
+func (n *LocalNode) AddValidators(validators ...*Validator) error {
+	n.Lock()
+	defer n.Unlock()
+
+	for _, va := range validators {
+		if n.Address() == va.Address() {
+			continue
+		}
+		n.validators[va.Address()] = va
+	}
+
+	return nil
+}
+
+func (n *LocalNode) RemoveValidators(validators ...*Validator) error {
+	n.Lock()
+	defer n.Unlock()
+
+	for _, va := range validators {
+		if _, ok := n.validators[va.Address()]; !ok {
+			continue
+		}
+		delete(n.validators, va.Address())
+	}
+
+	return nil
+}
+
+func (n *LocalNode) MarshalJSON() ([]byte, error) {
+	var neighbors = make(map[string]struct{})
+	for _, neighbor := range n.validators {
+		neighbors[neighbor.Address()] = struct{}{}
+	}
+	return json.Marshal(map[string]interface{}{
+		"address":    n.Address(),
+		"alias":      n.Alias(),
+		"endpoint":   n.Endpoint().String(),
+		"state":      n.State().String(),
+		"validators": neighbors,
```

Comment:
`json.Marshal()` also calls `sebaknode.Validator.MarshalJSON()`, so the `neighbors` is not necessary in here.

Reply:
> @Charleslee522 : You mean, no need 148 line right? I cannot understand.

`Validator` also have `MarshalJSON()` method, so we can just use without making `neighbors`
```
"validators": n.validators,
```
I did not tested :)

Reply:
@spikeekips You are right and it is more accurate to make localNode to JSON